src/Esprima/Ast/List.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Esprima.Ast
+{
+    // This structure is like List<> from the BCL except it is designed
+    // to be modifiable by this library during the construction of the AST
+    // but publicly read-only thereafter. The only allocation required on the
+    // heap is the backing array storage for the element. An empty list,
+    // however causes no heap allocation; that is, the array is allocated
+    // on first addition.
+
+    public struct List<T> : IReadOnlyList<T>
+    {
+        T[] _items;
+
+        public List(int capacity)
```

Comment:
should there be explicit no-arg constructor and/or otherwise detect 0 capacity => Array.Empty<T>?

Reply:
> should there be explicit no-arg constructor

A `struct` always has one.

This constructor should in fact be internal because it doesn't make sense for anyone else because they cannot do anything with that initial capacity. I'll fix this.

Reply:
Ah, sorry I misunderstood you and that's a fair point. I'll fix that.

Reply:
Addressed with 3ad476bd9ca527da712f549c18dc39d45a03a58f.